### ISSUE TITLE
May swiper-isearch provide an option not showing multiple times for one line's multiple match?

### ISSUE THREAD

**Issue by dolmens** [abo-abo/swiper#2034]
I personally still want to use avy to move to the match quickly, and dislike the duplicated lines for multiple matches in one line, It's somewhat not that beautiful as other ivy things.

**Comment by otadmor** [comment]
One appriach cold be count-lines between two search forward.
Second approach would be to search forward a \n after each existing search forward. Searching incrementally for \n could be a fast way of getting line numbers with only one pass on the buffer.

**Comment by basil-conto** [comment]
@dolmens Why wouldn't you use `swiper` instead of `swiper-isearch` for that?

**Comment by basil-conto** [comment]
@otadmor FWIW, `count-lines` uses the current buffer's newline cache, so it shouldn't be too slow.

**Comment by otadmor** [comment]
I have actually benchmarked line-number-at-pos and found it very slow. As line-number-at-pos uses count-lines I assumed count-lines was the slow one.
Does the "buffer's newline cache" expects the whole buffer to be visited in the past? This might not be a good assumption when using swiper.

https://github.com/emacs-mirror/emacs/blob/master/lisp/simple.el

**Comment by basil-conto** [comment]
> I have actually benchmarked line-number-at-pos and found it very slow. As line-number-at-pos uses count-lines I assumed count-lines was the slow one.

Yes, `count-lines` is probably the lion's share of `line-number-at-pos`.  I didn't say `count-lines` is negligible; I said it shouldn't be too slow when called once for each change in `swiper`. Having said that, I have not understood what you are suggesting to do with `count-lines`, and there is likely a better way to code this, as calling `count-lines` too frequently is an anti-pattern.

> Does the "buffer's newline cache" expects the whole buffer to be visited in the past?

Sorry, I don't know.

**Comment by otadmor** [comment]
I don't really know whether that would help, but instead of "count-lines" we could do something like (search-forward (concat "\n|" ivy--re) nil 'noting )
And then decide if we want to count a new line depends on looking-at. I'm not sure if this is faster as this would cause many elisp - native c jumps. This is probably depends on the buffer s content.

**Comment by basil-conto** [comment]
@otadmor Can you please explain what the purpose of doing that is? I don't understand how that relates to the OP.

**Comment by otadmor** [comment]
New candidates should not be pushed unless a new \n was found. This would resolve this issue.

**Comment by basil-conto** [comment]
How would that be different to using `swiper` instead of `swiper-isearch`?

**Comment by otadmor** [comment]
They could give different performance.

**Comment by basil-conto** [comment]
> Swiper saves the whole buffer in its own structure (list) and the could take time.

AFAICT, `swiper-isearch` does the same. In other words, it's not incremental.

**Comment by otadmor** [comment]
swiper-async does (re-search-forward re)
With given
[Code]
  (let* ((re-full (funcall ivy--regex-function str))
         (re (ivy-re-to-str re-full))
[/Code]

Then it does buffer-substring with the line caused re-search-forward to return. 
The swiper-isearch will return all the lines in the buffer only if re will match a newline.

**Comment by basil-conto** [comment]
I'll answer my own question: the difference between `swiper` and restricting `swiper-isearch` to a single match per line is that the former collects all lines in the buffer before Ivy applies its filtering, whereas the latter only collects those lines which match the current input. There is a tradeoff between collecting all buffer lines only once, and collecting them every time the input changes.

The OP was about limiting matches to one per line, which is what `swiper` already does. The OP did not say that `swiper` was so slow to start up that they had to use `swiper-isearch` instead. That would be a separate issue, I think.

**Comment by otadmor** [comment]
The question of using swiper or swiper-isearch is indeed irrelevant here, its the author choice to use swiper-isearch. I agree the suggestion of using swiper is the quickest solution which does not require any coding. As of implementing this feature for swiper-isearch - it is a nice discussion of how to do it the best and I believe everyone will benefit of it - both us and who-ever would want to implement it in the future.
BTW - I believe the second quickest solution require about one line of code with beginning-of-line/forward-line after finding a match.

**Comment by basil-conto** [comment]
> I believe the second quickest solution require about one line of code with beginning-of-line/forward-line after finding a match.

Would you be willing to provide a PR for this?

**Comment by otadmor** [comment]
No

**Comment by abo-abo** [comment]
@dolmens I agree that it's not beautiful. But I don't want to break the invariants of `ivy`. If there are 6 candidates in the collection, there must be 6 candidates in the minibuffer. Otherwise, the whole logic breaks.

```
cou cou
counsel.el (counsel-company): Use company-prefix
```

For the input `co`, `swiper-isearch` has 6 candidates, and `swiper` has 2 candidates. <kbd>C-n</kbd> and <kbd>C-p</kbd> can move between them, they don't have to know that they're dealing with either `swiper` or `swiper-isearch`: they're just generic `ivy` functions.

What we can do is fix the highlighting. Right now, the highlights look like this:

```
|co|u |co|u
|co|u |co|u
|co|unsel.el (|co|unsel-|co|mpany): Use |co|mpany-prefix
|co|unsel.el (|co|unsel-|co|mpany): Use |co|mpany-prefix
|co|unsel.el (|co|unsel-|co|mpany): Use |co|mpany-prefix
|co|unsel.el (|co|unsel-|co|mpany): Use |co|mpany-prefix
```

It would look at lot better like this:

```
|co|u cou
cou |co|u
|co|unsel.el (counsel-company): Use company-prefix
counsel.el (|co|unsel-company): Use company-prefix
counsel.el (counsel-|co|mpany): Use company-prefix
counsel.el (counsel-company): Use |co|mpany-prefix
```

I'll try to implement this when I get some time.

**Comment by abo-abo** [comment]
I just got the idea that it can also be improved in `ivy-format-function` without breaking the invariants for <kbd>C-n</kbd> / <kbd>C-p</kbd>.
This function gets a list of e.g. 10 candidates to display in the minibuffer. It can easily detect consecutive strings that are the same and fold them together in a way that still shows which one of the folded equal strings is selected.

**Comment by basil-conto** [comment]
> This function gets a list of e.g. 10 candidates to display in the minibuffer. It can easily detect consecutive strings that are the same and fold them together in a way that still shows which one of the folded equal strings is selected.

Will this work even when one of the consecutive candidates is scrolled out of view?

**Comment by abo-abo** [comment]
Thanks, all. I think it looks a lot better now. Please test.

**Comment by manuel-uberti** [comment]
@abo-abo is something changed after your fix? I'm not able to theme `swiper-isearch-current-match`, see for instance https://github.com/purcell/color-theme-sanityinc-tomorrow/pull/131

**Comment by abo-abo** [comment]
@manuel-uberti That face is now removed, see the comment on PR. You can style `swiper-line-face` instead.
